src/languages/reasonml.js

```diff
@@ -6,301 +6,63 @@ Author: Gidi Meir Morris <oss@gidi.io>
-    relevance: 0,
-    contains: [
-      CONSTRUCTOR_MODE,
-      OPERATOR_MODE,
-      {
-        relevance: 0,
-        className: 'constructor',
-        begin: RE_MODULE_IDENT
-      }
-    ]
-  };
-
-  const MODULE_ACCESS_MODE = {
-    className: 'module-access',
-    keywords: KEYWORDS,
-    returnBegin: true,
-    variants: [
-      { begin: "\\b(" + RE_MODULE_IDENT + "\\.)+" + RE_IDENT },
-      {
-        begin: "\\b(" + RE_MODULE_IDENT + "\\.)+\\(",
-        end: "\\)",
-        returnBegin: true,
-        contains: [
-          FUNCTION_BLOCK_MODE,
-          {
-            begin: '\\(',
-            end: '\\)',
-            relevance: 0,
-            skip: true
-          }
-        ].concat(MODULE_ACCESS_CONTENTS)
-      },
-      {
-        begin: "\\b(" + RE_MODULE_IDENT + "\\.)+\\{",
-        end: /\}/
-      }
-    ],
-    contains: MODULE_ACCESS_CONTENTS
-  };
-
-  PARAMS_CONTENTS.push(MODULE_ACCESS_MODE);
-
   return {
     name: 'ReasonML',
     aliases: [ 're' ],
-    keywords: KEYWORDS,
+    keywords: {
+      $pattern: '[a-z_]\\w*!?',
+      keyword:
+        'and as asr assert begin class constraint do done downto else end esfun '
+        + 'exception external for fun function functor if in include inherit '
+        + 'initializer land lazy let lor lsl lsr lxor mod module mutable new '
+        + 'nonrec object of open or pri pub rec sig struct switch then to try '
+        + 'type val virtual when while with',
+      built_in:
+        'array bool bytes char exn|5 float int int32 int64 list lazy_t|5 nativeint|5 ref string unit ',
+      literal:
+        'true false'
+    },
     illegal: '(:-|:=|\\$\\{|\\+=)',
     contains: [
-      hljs.COMMENT('/\\*', '\\*/', { illegal: '^(#,\\/\\/)' }),
-      {
-        className: 'character',
-        begin: '\'(\\\\[^\']+|[^\'])\'',
-        illegal: '\\n',
-        relevance: 0
-      },
-      hljs.QUOTE_STRING_MODE,
       {
         className: 'literal',
-        begin: '\\(\\)',
+        begin: '\\[(\\|\\|)?\\]|\\(\\)',
         relevance: 0
       },
-      {
-        className: 'literal',
-        begin: '\\[\\|',
-        end: '\\|\\]',
-        relevance: 0,
-        contains: LIST_CONTENTS_MODES
+      hljs.C_LINE_COMMENT_MODE,
+      hljs.COMMENT('/\\*', '\\*/', { illegal: '^(#,\\/\\/)' }),
+      { /* type variable */
+        className: 'symbol',
+        begin: '\'[A-Za-z_](?!\')[\\w\']*'
+        /* the grammar is ambiguous on how 'a'b should be interpreted but not the compiler */
       },
-      {
-        className: 'literal',
-        begin: '\\[',
-        end: '\\]',
-        relevance: 0,
-        contains: LIST_CONTENTS_MODES
+      { /* polymorphic variant */
+        className: 'type',
+        begin: '`[A-Z][\\w\']*'
       },
-      CONSTRUCTOR_MODE,
-      {
-        className: 'operator',
-        begin: RE_OPERATOR_SPACED,
-        illegal: '-->',
+      { /* module or constructor */
+        className: 'type',
+        begin: '\\b[A-Z][\\w\']*',
```

Comment:
If there is no `end` matcher then you should use `match` instead of `begin` for clarity.

Reply:
https://github.com/highlightjs/highlight.js/pull/3795/commits/7f04ca8b37f494d5b6999b3b77a815f59880671d